### ISSUE TITLE
onStart called before activation criteria is met

### ISSUE THREAD

**Issue by mrousavy** [software-mansion/react-native-gesture-handler#1225]
## TL;DR

* On iOS `onStart` and `onEnd` are both only called when the activation criteria is met (e.g. X `[-5, 5]`)
* On Android `onStart` is called before activation criteria is met, and `onEnd` is not, meaning anything that depends on an `onEnd` "cleanup" function won't work on Android.

## Description

I'm having a horizontal image swiper implemented using a `<PanGestureHandler>` which does the following logic:

1. `onStart`: Cancels any animations on the translateX animation
2. `onActive`: Sets the translateX value to the gesture handler's state translation.x value
3. `onEnd`: Runs a spring animation on the translateX value to snap to an edge

I've set the activation criteria to the following:

```jsx
<PanGestureHandler
  ref={horizontalGestureHandler}
  enabled={enableHorizontalSwipe}
  onGestureEvent={onHorizontalGestureEvent}
  activeOffsetX={HORIZONTAL_SWIPE_ACTIVE_X}
  failOffsetY={HORIZONTAL_SWIPE_FAIL_Y}
  maxPointers={1}>
```

Then I've added `console.log`s in onStart, onActive and onEnd.

When the user starts dragging with velocity and then releases the finger, you'd expect `onEnd` to take over and run the spring animation to snap to an edge. While this animation is running, the user can activate the gesture again to cancel the animation and start taking over swiping again. 

When cancelling the `onEnd` with a gesture (i.e. dragging while the snap to edge spring animation is still playing), the user can also just _tap_ the screen, which results in two different cases on the platforms:

* iOS: Nothing is called until the activation criteria is met, in my case [-5, 5] on X, so the animation continues to play until the user actually starts dragging instead of tapping
* Android: `onStart` is called, even if the activation criteria isn't met yet (pointer hasn't moved [-5, 5] on X yet), and on release `onEnd` is not called, because activation criteria wasn't met! (Note: `onFinish` will be called, but I can't use that since then cross-handler interaction is weird)

In other words: on iOS everything works fine, but on Android `onStart` is called even before the activation criteria is met, causing the swiper to get stuck where it won't start the `onEnd` snap-to-edges spring animation.

### Demo

* [iOS](https://i.imgur.com/SHH35IG.mp4)
* [Android](https://i.imgur.com/biZI6A1.mp4)

> What you're seeing is me spamming my left mouse button, the "finger" (my mouse) hasn't moved a single pixel in this demo, and should therefore not result in `onStart` being called since that requires the activation criteria of 5 pixel to the left or 5 pixel to the right.

## Solutions

Only call `onStart` on Android if the activation criteria is actually met!

## Package versions

- React: 16
- React Native: 0.63.3
- React Native Gesture Handler: 1.8.0
- React Native Reanimated: 2.0.0-alpha.8

**Comment by mym0404** [comment]
+1

**Comment by balazs31** [comment]
+1

Can confirm the same behavior on Android: `onStart` is fired if I long tap on the item without swiping, but `onEnd` won't be fired upon release.

 ```
 <PanGestureHandler
  activeOffsetX={[-10, 10]}
  onGestureEvent={onGestureEvent}>
  {renderContent()}
 </PanGestureHandler>
```

**Comment by hubciorz** [comment]
Any workaround for that? It's really annoying if you have multiple nested handlers :(

**Comment by intergalacticspacehighway** [comment]
Tried the below snippet and couldn't reproduce it.

```jsx
const handler = useAnimatedGestureHandler({
    onStart(e) {
      console.log("start ", e.x);
    },
    onActive(e) {
      console.log("active ", e.x);
    },
    onEnd(e) {
      console.log("end ", e.x);
    },
    onFail(e) {
      console.log("failed ", e.x);
    },
    onCancel(e) {
      console.log("cancel ", e.x);
    },
    onFinish(e) {
      console.log("finish ", e.x);
    },
  });
```

```jsx
<PanGestureHandler
    onGestureEvent={handler}
    activeOffsetX={[-200, 200]}
    maxPointers={1}
>
    <Animated.View
      style={{ height: 300, width: 300, backgroundColor: "pink" }}
    />
</PanGestureHandler>
```

For me, `onStart` gets called as soon as a finger is tapped (on iOS and Android). When I release finger, `onFail` and `onFinish` gets called if the criteria of `activeOffsetX` is not met. This happens consistently for me on iOS and Android. `onEnd` is not called on both platforms unless `onActive` is called (i.e. criteria is met). Maybe this issue was fixed. 🤔

**Comment by j-piasecki** [comment]
Hi! As @intergalacticspacehighway mentioned, this issue should be fixed by Gesture Handler 2.0. Besides some changes to native handlers to make state flow more consistent across platforms, the new API introduced a new callback: `onBegin`, which is called whenever the gesture starts processing touch stream (you can use `onFinalize` to clean up after it). `onStart` is now called whenever the gesture activates (and you can use `onEnd` to clean up after it).
